src/libraries/System.Web.HttpUtility/src/System/Web/Util/UriUtil.cs

```diff
@@ -10,35 +10,36 @@ internal static class UriUtil
         // Just extracts the query string and fragment from the input path by splitting on the separator characters.
         // Doesn't perform any validation as to whether the input represents a valid URL.
         // Concatenating the pieces back together will form the original input string.
-        private static void ExtractQueryAndFragment(string input, out string path, out string? queryAndFragment)
+        private static void ExtractQueryAndFragment(string input, out Range pathRange, out Range queryAndFragmentRange)
         {
             int queryFragmentSeparatorPos = input.AsSpan().IndexOfAny('?', '#'); // query fragment separators
             if (queryFragmentSeparatorPos >= 0)
             {
-                path = input.Substring(0, queryFragmentSeparatorPos);
-                queryAndFragment = input.Substring(queryFragmentSeparatorPos);
+                pathRange = new Range(Index.Start, Index.FromStart(queryFragmentSeparatorPos));
+                queryAndFragmentRange = new Range(Index.FromStart(queryFragmentSeparatorPos), Index.End);
             }
             else
             {
                 // no query or fragment separator
-                path = input;
-                queryAndFragment = null;
+                pathRange = Range.All;
+                queryAndFragmentRange = default(Range);
             }
         }
 
         // Attempts to split a URI into its constituent pieces.
         // Even if this method returns true, one or more of the out parameters might contain a null or empty string, e.g. if there is no query / fragment.
         // Concatenating the pieces back together will form the original input string.
-        internal static bool TrySplitUriForPathEncode(string input, [NotNullWhen(true)] out string? schemeAndAuthority, [NotNullWhen(true)] out string? path, out string? queryAndFragment)
+        internal static bool TrySplitUriForPathEncode(string input, out Range schemeAndAuthority, out Range path, out Range queryAndFragment)
```

Comment:
By forcing everything into spans and ranges, you are now also allocating substrings where you might not have needed to previously. E.g. if `pathRange` is `Range.All`, you're reallocating the original string.

You could inline `ExtractQueryAndFragment` into this method, and always use `Substring` for the path portion.
I would also suggest returning `Span` instead of `Range` as it's easier to use given that the callers only need the range to make the span anyway.